docs/man/man8/badmin.8
.ds ]W %
.ds ]L
.nh
.TH badmin 8 "Lava Version 1.0 - Sept 2007"
.br
.SH NAME
\fBbadmin\fR - administrative tool for Lava
.SH SYNOPSIS
.BR
.PP
.PP
\fBbadmin\fR\fB \fR\fIsubcommand
\fR.PP
\fBbadmin\fR\fB \fR[\fB-h\fR | \fB-V\fR]
.SH SUBCOMMAND LIST
.BR
.PP
.br
\fBckconfig\fR [\fB-v\fR]
.br
\fBreconfig\fR [\fB-v\fR] [\fB-f\fR]
.br
\fBmbdrestart\fR [\fB-v\fR] [\fB-f\fR]
.br
\fBqopen\fR [\fIqueue_name\fR ... | \fBall\fR]
.br
\fBqclose\fR [\fIqueue_name\fR ... | \fBall\fR]
.br
\fBqact\fR [\fIqueue_name\fR ... | \fBall\fR]
.br
\fBqinact\fR [\fIqueue_name\fR ... | \fBall\fR]
.br
\fBqhist\fR [\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR] [\fIqueue_name \fR...]
.br
\fBhopen\fR [\fIhost_name\fR ... | \fIhost_group\fR ... | \fBall\fR]
.br
\fBhclose\fR [\fIhost_name\fR ... | \fIhost_group\fR ... | \fBall\fR]
.br
\fBhrestart\fR [\fB-f\fR] [\fIhost_name\fR ...| \fBall\fR]
.br
\fBhshutdown\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBhstartup\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBhhist\fR [\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR] [\fIhost_name\fR ...]
.br
\fBmbdhist \fR[\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR]
.br
\fBhist\fR [\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR]
.br
\fBhelp\fR [\fIcommand\fR ...] | \fB?\fR [\fIcommand\fR ...]
.br
\fBquit
\fR.br
\fBsbddebug\fR [\fB-c\fR \fIclass_name ...\fR] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] 
[\fIhost_name ...\fR]
.br
\fBmbddebug\fR [\fB-c\fR \fIclass_name ...\fR] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR]
.br
\fBsbdtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fIhost_name ...\fR]
.br
\fBmbdtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name]\fR [\fB-o\fR]
.SH DESCRIPTION
.BR
.PP
.PP
This command can only be used by Lava administrators.
.PP
badmin provides a set of commands to control and monitor Lava. If no 
subcommands are supplied for badmin, badmin prompts for a 
command from standard input. Commands bqc(8), breconfig(8) and 
breboot(8) are superceded by badmin(8).
.PP
Information about each command is available through the help 
command.
.PP
The badmin commands consist of a set of privileged commands and a 
set of non\fB-\fRprivileged commands. Privileged commands can only be 
invoked by root or Lava administrators as defined in the configuration 
file (see lsf.cluster.cluster(5) for ClusterAdmin). Privileged 
commands are:
.PP
reconfig
.PP
mbdrestart
.PP
qopen
.PP
qclose
.PP
qact
.PP
qinact
.PP
hopen
.PP
hclose
.PP
hrestart
.PP
hshutdown
.PP
hstartup
.PP
All other commands are non\fB-\fRprivileged commands and can be invoked 
by any Lava user. If the privileged commands are to be executed by an 
Lava administrator, badmin must be installed setuid root, because it 
needs to send the request using a privileged port.
.PP
For subcommands for which multiple host names or host groups can 
be specified, do not enclose the multiple names in quotation marks. 
.SH OPTIONS
.BR
.PP
.TP 
\fIsubcommand
\fR
.IP
Executes the specified subcommand. See Usage section.


.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits.


.TP 
\fB-V
\fR
.IP
Prints Lava release version to stderr and exits.


.SH USAGE
.BR
.PP
.TP 
\fBckconfig\fR [\fB-v\fR]

.IP
Checks Lava configuration files. Configuration files are located in the 
LSB_CONFDIR/\fIcluster_name\fR/configdir directory. 

.IP
The LSB_CONFDIR variable is defined in lsf.conf (see lsf.conf(5)) 
which is in LSF_ENVDIR or /etc (if LSF_ENVDIR is not defined).

.IP
By default, badmin ckconfig displays only the result of the 
configuration file check. If warning errors are found, badmin prompts 
you to display detailed messages.


.IP
\fB-v\fR
.BR
.RS
.IP
Verbose mode. Displays detailed messages about configuration 
file checking to stderr.

.RE

.TP 
\fBreconfig\fR [\fB-v\fR] [\fB-f\fR]

.IP
Dynamically reconfigures Lava without restarting MBD. Configuration 
files are checked for errors and the results displayed to stderr. If no 
errors are found in the configuration files, a reconfiguration request is 
sent to MBD and configuration files are reloaded.

.IP
With this command, MBD is not restarted and lsb.events is not 
replayed. To restart MBD and replay lsb.events, use badmin 
mbdrestart.

.IP
When you issue this command, MBD is available to service requests 
while reconfiguration files are reloaded. Configuration changes made 
since system boot or the last reconfiguration take effect.

.IP
If warning errors are found, badmin prompts you to display detailed 
messages. If fatal errors are found, reconfiguration is not performed, 
and badmin exits.

.IP
If you add a host to a host group, or a host to a queue, the new host 
will not be recognized by jobs that were submitted before you 
reconfigured. If you want the new host to be recognized, you must use 
the command badmin mbdrestart.


.IP
\fB-v\fR
.BR
.RS
.IP
Verbose mode. Displays detailed messages about the status of 
the configuration files. Without this option, the default is to 
display the results of configuration file checking. All messages 
from the configuration file check are printed to stderr.

.RE

.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and proceeds with reconfiguration if 
configuration files contain no fatal errors.

.RE

.TP 
\fBmbdrestart\fR [\fB-v\fR] [\fB-f\fR]

.IP
Dynamically reconfigures Lava and restarts MBD. Configuration files are 
checked for errors and the results printed to stderr. If no errors are 
found, configuration files are reloaded, MBD is restarted, and events in 
lsb.events are replayed to recover the running state of the last MBD. 
MBD is unavailable to service requests while it restarts.

.IP
If warning errors are found, badmin prompts you to display detailed 
messages. If fatal errors are found, MBD restart is not performed, and 
badmin exits.

.IP
If lsb.events is large, or many jobs are running, restarting MBD can 
take several minutes. If you only need to reload the configuration files, 
use badmin reconfig.


.IP
\fB-v\fR 
.BR
.RS
.IP
Verbose mode. Displays detailed messages about the status of 
configuration files. All messages from configuration checking 
are printed to stderr.

.RE

.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and forces reconfiguration and MBD restart 
to proceed if configuration files contain no fatal errors.

.RE

.TP 
\fBqopen\fR [\fIqueue_name ... \fR| \fBall\fR]

.IP
Opens specified queues, or all queues if the reserved word all is 
specified. If no queue is specified, the system default queue is assumed 
(see lsb.queues(5) for DEFAULT_QUEUE). A queue can accept batch 
jobs only if it is open.


.TP 
\fBqclose\fR [\fIqueue_name\fR ... | \fBall\fR]

.IP
Closes specified queues, or all queues if the reserved word all is 
specified. If no queue is specified, the system default queue is 
assumed. A queue will not accept any job if it is closed.


.TP 
\fBqact \fR[\fIqueue_name\fR ... | \fBall\fR]

.IP
Activates specified queues, or all queues if the reserved word all is 
specified. If no queue is specified, the system default queue is 
assumed. Jobs in a queue can be dispatched if the queue is activated. 

.IP
A queue inactivated by its run windows cannot be reactivated by this 
command (see lsb.queues(5) for RUN_WINDOW).


.TP 
\fBqinact\fR [\fIqueue_name\fR ... | \fBall\fR]

.IP
Inactivates specified queues, or all queues if the reserved word all is 
specified. If no queue is specified, the system default queue is 
assumed. No job in a queue can be dispatched if the queue is 
inactivated.


.TP 
\fBqhist\fR [\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR] [\fIqueue_name\fR ...]

.IP
Displays historical events for specified queues, or for all queues if no 
queue is specified. Queue events are queue opening, closing, 
activating and inactivating.


.IP
\fB-t\fR \fItime0\fR,\fItime1\fR
.BR
.RS
.IP
Displays only those events that occurred during the period 
from \fItime0\fR to \fItime1\fR. See bhist(1) for the time format. The 
default is to display all queue events in the event log file (see 
below).

.RE

.IP
\fB-f\fR \fIlogfile_name\fR
.BR
.RS
.IP
Specify the file name of the event log file. Either an absolute or 
a relative path name may be specified. The default is to use the 
event log file currently used by the Lava system: 
LSB_SHAREDIR/cluster_name/logdir/lsb.events. Option 
-f is useful for offline analysis.

.RE

.TP 
\fBhopen\fR [\fIhost_name \fR... | \fIhost_group \fR... | \fBall\fR]

.IP
Opens batch server hosts. Specify the names of any server hosts or host 
groups (see bmgroup(1)). All batch server hosts will be opened if the 
reserved word all is specified. If no host or host group is specified, 
the local host is assumed. A host accepts batch jobs if it is open.


.TP 
\fBhclose\fR [\fIhost_name \fR... | \fIhost_group \fR... | \fBall\fR]

.IP
Closes batch server hosts. Specify the names of any server hosts or host 
groups (see bmgroup(1)). All batch server hosts will be closed if the 
reserved word all is specified. If no argument is specified, the local 
host is assumed. A closed host will not accept any new job, but jobs 
already dispatched to the host will not be affected. Note that this is 
different from a host closed by a window \fB-\fR all jobs on it are suspended 
in that case.


.TP 
\fBhrestart\fR [\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Restarts SBD on the specified hosts, or on all server hosts if the 
reserved word all is specified. If no host is specified, the local host is 
assumed. SBD will re\fB-\fRexecute itself from the beginning. This allows 
new SBD binaries to be used.


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
restarting SBDs. 

.RE

.TP 
\fBhshutdown\fR [\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Shuts down SBD on the specified hosts, or on all batch server hosts if 
the reserved word all is specified. If no host is specified, the local host 
is assumed. SBD will exit upon receiving the request.


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
shutting down SBDs. 

.RE

.TP 
\fBhstartup\fR [\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Starts up SBD on the specified hosts, or on all batch server hosts if the 
reserved word all is specified. Only root 
can use this option, and those users must be able to 
use rsh on all Lava hosts. If no host is specified, the local host is 
assumed.


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
starting up SBDs. 

.RE

.TP 
\fBhhist\fR [\fB-t\fR \fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR] [\fIhost_name \fR...]

.IP
Displays historical events for specified hosts, or for all hosts if no host 
is specified. Host events are host opening and closing. Options -t and 
-f are exactly the same as those of qhist (see above).


.TP 
\fBmbdhist\fR [\fB-t \fR\fItime0\fR,\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR]

.IP
Displays historical events for MBD. Events describe the starting and 
exiting of MBD. Options -t and -f are exactly the same as those of 
qhist (see above).


.TP 
\fBhist\fR [\fB-t\fR \fItime0\fR\fI,\fR\fItime1\fR] [\fB-f\fR \fIlogfile_name\fR]

.IP
Displays historical events for all the queues, hosts and MBD. Options 
-t and -f are exactly the same as those of qhist (see above).


.TP 
\fBhelp\fR [\fIcommand\fR\fI ...\fR] | \fB?\fR [\fIcommand\fR\fI ...\fR]

.IP
Displays the syntax and functionality of the specified commands.


.TP 
\fBquit
\fR
.IP
Exits the badmin session.


.TP 
\fBsbddebug\fR [\fB-c\fR \fIclass_name ...\fR] [\fB-l\fR \fIdebug_level\fR] [\fB-\fR\fBf\fR \fIlogfile_name\fR] [\fB-o\fR] 
[\fIhost_name ...\fR]

.IP
Sets the message log level for SBD to include additional information in 
log files. You must be root or the Lava administrator to use this 
command.

.IP
If the command is used without any options, the following default 
values are used:

.IP
\fIclass_name\fR = 0 (no additional classes are logged)

.IP
\fIdebug_level\fR = 0 (LOG_DEBUG level in parameter LSF_LOG_MASK)

.IP
\fIlogfile_name\fR = current Lava system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name\fR.log.\fIhost_name
\fR
.IP
\fIhost_name\fR = local host (host from which command was submitted)


.IP
\fB-c\fR \fIclass_name ...\fR
.BR
.RS
.IP
Specifies software classes for which debug messages are to be 
logged.

.IP
Format of \fIclass_name \fRis the name of a class, or a list of class 
names separated by spaces and enclosed in quotation marks.

.IP
Possible classes:

.IP
LC_AUTH - Log authentication messages

.IP
LC_CHKPNT - Log checkpointing messages

.IP
LC_COMM - Log communication messages

.IP
LC_EXEC - Log significant steps for job execution

.IP
LC_FILE - Log file transfer messages

.IP
LC_HANG - Mark where a program might hang

.IP
LC_JLIMIT - Log job slot limit messages

.IP
LC_LOADINDX - Log load index messages

.IP
LC_PEND - Log messages related to job pending reasons

.IP
LC_PERFM - Log performance messages

.IP
LC_PIM - Log PIM messages

.IP
LC_SIGNAL - Log messages pertaining to signals

.IP
LC_SYS - Log system call messages

.IP
LC_TRACE - Log significant program walk steps

.IP
LC_XDR - Log everything transferred by XDR

.IP
Note: Classes are also listed in lsf.h.

.IP
Default: 0 (no additional classes are logged)

.RE

.IP
\fB-l\fR \fIdebug_level\fR
.BR
.RS
.IP
Specifies level of detail in debug messages. The higher the 
number, the more detail that is logged. Higher levels include all 
lower levels.

.IP
Possible values:

.IP
0 LOG_DEBUG level in parameter LSF_LOG_MASK in 
lsf.conf. 

.IP
1 LOG_DEBUG1 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2 LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
2 LOG_DEBUG2 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2 LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
3 LOG_DEBUG3 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2, LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
Default: 0 (LOG_DEBUG level in parameter LSF_LOG_MASK)

.RE

.IP
\fB-f\fR \fIlogfile_name\fR
.BR
.RS
.IP
Specify the name of the file into which debugging messages are 
to be logged. A file name with or without a full path may be 
specified.

.IP
If a file name without a path is specified, the file will be saved 
in the directory indicated by the parameter LSF_LOGDIR in 
lsf.conf.

.IP
The name of the file that will be created will have the following 
format:

.IP
\fIlogfile_name.daemon_name.\fRlog\fI.host_name
\fR
.IP
If the specified path is invalid, on UNIX, the log file is created 
in the /tmp directory. 

.IP
If LSF_LOGDIR is not defined, daemons log to the syslog 
facility.

.IP
Default: current Lava system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name\fR.log\fI.host_name\fR.

.RE

.IP
\fB-o\fR
.BR
.RS
.IP
Turns off temporary debug settings and resets them to the 
daemon starting state. The message log level is reset back to the 
value of LSF_LOG_MASK and classes are reset to the value of 
LSB_DEBUG_MBD, LSB_DEBUG_SBD.

.IP
The log file is also reset back to the default log file.

.RE

.IP
\fIhost_name ...\fR
.BR
.RS
.IP
Optional. Sets debug settings on the specified host or hosts.

.IP
Lists of host names must be separated by spaces and enclosed 
in quotation marks.

.IP
Default: local host (host from which command was submitted)

.RE

.TP 
\fBmbddebug\fR [\fB-c\fR \fIclass_name ...\fR] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR]

.IP
Sets message log level for MBD to include additional information in log 
files. You must be root or the Lava administrator to use this command.

.IP
See sbddebug for an explanation of options.


.TP 
\fBsbdtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fIhost_name ...\fR]

.IP
Sets the timing level for SBD to include additional timing information 
in log files. You must be root or the Lava administrator to use this 
command.

.IP
If the command is used without any options, the following default 
values are used:

.IP
\fItiming_level\fR = no timing information is recorded

.IP
\fIlogfile_name\fR = current Lava system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name.\fRlog\fI.host_name
\fR
.IP
\fIhost_name \fR= local host (host from which command was submitted)


.IP
\fB-l \fR\fItiming_level\fR
.BR
.RS
.IP
Specifies detail of timing information that is included in log 
files. Timing messages indicate the execution time of functions 
in the software and are logged in milliseconds.

.IP
Valid values: 1 | 2 | 3 | 4 | 5

.IP
The higher the number, the more functions in the software that 
are timed and whose execution time is logged. The lower 
numbers include more common software functions. Higher 
levels include all lower levels.

.IP
Default: undefined (no timing information is logged)

.RE

.IP
\fB-f\fR \fIlogfile_name\fR
.BR
.RS
.IP
Specify the name of the file into which timing messages are to 
be logged. A file name with or without a full path may be 
specified.

.IP
If a file name without a path is specified, the file will be saved 
in the directory indicated by the parameter LSF_LOGDIR in 
lsf.conf.

.IP
The name of the file that will be created will have the following 
format:

.IP
\fIlogfile_name.daemon_name.\fRlog\fI.host_name
\fR
.IP
If the specified path is invalid, on UNIX, the log file is created 
in the /tmp directory.

.IP
If LSF_LOGDIR is not defined, daemons log to the syslog 
facility.

.IP
\fBNote: \fRBoth timing and debug messages are logged in the same 
files.

.IP
Default: current Lava system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name.\fRlog\fI.host_name\fR.

.RE

.IP
\fB-o\fR
.BR
.RS
.IP
Optional. Turn off temporary timing settings and reset them to 
the daemon starting state. The timing level is reset back to the 
value of the parameter for the corresponding daemon 
(LSB_TIME_MBD, LSB_TIME_SBD).

.IP
The log file is also reset back to the default log file.

.RE

.IP
\fIhost_name \fR...
.BR
.RS
.IP
Sets the timing level on the specified host or hosts.

.IP
Lists of hosts must be separated by spaces and enclosed in 
quotation marks.

.IP
Default: local host (host from which command was submitted)

.RE

.TP 
\fBmbdtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR]

.IP
Sets timing level for MBD to include additional timing information in 
log files. You must be root or the Lava administrator to use this 
command.

.IP
See sbdtime for an explanation of options.


.SH SEE ALSO
.BR
.PP
.PP
bqueues(1), bhosts(1), lsb.queues(5), lsb.hosts(5), 
lsf.conf(5), lsf.cluster(5), sbatchd(8), mbatchd(8)


src/lsbatch/cmd/badmin.c
/* $Id: badmin.c,v 1.8 2007/08/15 22:18:43 tmizan Exp $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsbatch/cmd/badmin.h"

extern char *myGetOpt (int nargc, char **nargv, char *ostr);
extern int checkConf (int, int);
extern int getConfirm (char *msg);
extern int lsb_debugReq(struct debugReq *pdebug , char *host);
static int doBatchCmd (int argc, char *argv[]);
static int badminDebug (int nargc, char *nargv[], int opCode);

int
main (int argc, char **argv)
{
    int cc,  myIndex;
    char *prompt = "badmin>";
    static char line[MAXLINELEN];

    if (lsb_init(argv[0]) < 0) {
	lsb_perror("lsb_init");
	exit(-1);
    }

    while ((cc = getopt(argc, argv, "Vh")) != EOF) {
        switch (cc) {
        case 'V':
            fprintf(stderr, "%s\n", LAVALITE_VERSION_STR);
            exit(0);
        case 'h':
        default:

            cmdsUsage("badmin", cmdList, cmdInfo );
        }
    }
    if (argc > optind) {


        if ((myIndex=adminCmdIndex(argv[optind], cmdList)) == -1) {
            fprintf(stderr, ("Invalid command <%s> \n"), argv[optind]);

            cmdsUsage("badmin", cmdList, cmdInfo);
	}
	optind++;
	int rc = doBatchCmd (argc, argv);
	exit ( rc );
    }

    for (;;) {
        printf("%s", prompt);
        fflush(stdout);
	if (fgets(line, MAXLINELEN, stdin) == NULL) {
	    printf("\n");
            exit(-1);
        }

        parseAndDo (line, doBatchCmd);
    }
    return(0);

}

static int
doBatchCmd (int argc, char *argv[])
{
    int cmdRet = 0, myIndex;

    if ((myIndex=adminCmdIndex(argv[optind-1], cmdList)) == -1) {
        fprintf(stderr,
                "Invalid command <%s>. Try help\n",
                argv[optind-1]);
        return -1;
    }
    switch (myIndex) {
    case BADMIN_MBDRESTART :
        cmdRet = breconfig (argc, argv, MBD_RESTART);
        break;
    case BADMIN_RECONFIG :
        cmdRet = breconfig (argc, argv, MBD_RECONFIG);
        break;
    case BADMIN_CKCONFIG :
        cmdRet = breconfig (argc, argv, MBD_CKCONFIG);
        break;
    case BADMIN_QOPEN :
    case BADMIN_QCLOSE :
    case BADMIN_QACT :
    case BADMIN_QINACT :
        cmdRet = bqc (argc, argv, opCodeList[myIndex]);
        break;
    case BADMIN_HOPEN :
    case BADMIN_HCLOSE :
    case BADMIN_HREBOOT :
    case BADMIN_HSHUT :
        cmdRet = bhc (argc, argv, opCodeList[myIndex]);
        break;
    case BADMIN_HSTARTUP :
        startup (argc, argv, myIndex);
	break;
    case BADMIN_QHIST:
    case BADMIN_HHIST:
    case BADMIN_MBDHIST:
    case BADMIN_HIST:
        cmdRet = sysHist(argc, argv, opCodeList[myIndex]);
        break;
    case BADMIN_MBDDEBUG:
    case BADMIN_MBDTIME:
    case BADMIN_SBDDEBUG:
    case BADMIN_SBDTIME:
        cmdRet = badminDebug(argc, argv, opCodeList[myIndex]);
        break;

    case BADMIN_HELP :
    case BADMIN_QES:
        cmdHelp (argc, argv, cmdList,
                 cmdInfo,
                 cmdSyntax);
        break;
    case BADMIN_QUIT:
        exit(0);
    default :
        fprintf(stderr, I18N_FUNC_S_ERROR, "adminCmdIndex()");
        exit(-1);
    }
    if (cmdRet == -2)
        oneCmdUsage(myIndex, cmdList, cmdSyntax);

    fflush(stderr);
    return (cmdRet);

}

int
breconfig (int argc, char **argv, int configFlag)
{
    extern int optind;
    char *optName;
    int vFlag = 0;
    int fFlag = 0;
    int checkReply;
    int stdoutsave;
    FILE *fp;
    char *linep;

    while ((optName = myGetOpt(argc, argv, "f|v|")) != NULL) {
	switch(optName[0]) {
        case 'v':
	    vFlag = 1;
	    break;
        case 'f':
            fFlag = 1;
            break;
	default:
	    return(-2);
	}
    }
    if ( optind < argc ) {
	return(-2);
    }

    if (!vFlag && !fFlag) {

        fprintf(stderr, ("\nChecking configuration files ...\n\n"));
        stdoutsave = dup(1);
        char tmpfile[] = "/tmp/badminXXXXXX";
        int fd = mkstemp(tmpfile);
        if (fd == -1) {
            perror("mkstemp");
            exit(1);
        }
        if (fd >= 0 ) {
            dup2(fd, 1);
            dup2(fd, 2);
            checkReply = checkConf(1, 2);
            close(fd);
            dup2(stdoutsave, 1);
            dup2(stdoutsave, 2);
            fp = fopen(tmpfile, "r");
            if (fp != 0) {
                if (checkReply == EXIT_FATAL_ERROR
                    || checkReply == EXIT_WARNING_ERROR) {
                    if (checkReply == EXIT_FATAL_ERROR)
                        fprintf(stderr, "There are fatal errors.\n\n");
                    else
                        fprintf(stderr, "There are warning errors.\n\n");
                    fflush(stderr);

                    if (getConfirm(("Do you want to see detailed messages? [y/n] ")))
                        while ((linep = getNextLine_(fp, 0)))
                            fprintf(stderr, "%s\n", linep);
                }
                else
                    fprintf(stderr, "No errors found.\n\n");
            }
            fclose(fp);
            unlink(tmpfile);
        }
        else
            checkReply = checkConf(0, 2);
    } else
        checkReply = checkConf(vFlag, 2);

    if (configFlag == MBD_CKCONFIG ) {
        return(0);
    }

    switch (checkReply)  {
    case EXIT_FATAL_ERROR :
        return -1;
    case EXIT_WARNING_ERROR :
        if (fFlag)
            break;
        if ( configFlag == MBD_RECONFIG ) {
            if (!getConfirm(("\nDo you want to reconfigure? [y/n] "))) {
                fprintf(stderr, ("Reconfiguration aborted.\n"));
                return(-1);
            }
	} else {
            if (!getConfirm("\nDo you want to restart MBD? [y/n] ")) {
                fprintf(stderr, ("MBD restart aborted.\n"));
                return(-1);
            }
        }
    default :
        ;
    }

    if (lsb_reconfig(configFlag) < 0) {
    	lsb_perror(("Failed"));
	return(-1);
    }

    if ( configFlag == MBD_RECONFIG ) {
        printf("%s\n",
               "Reconfiguration initiated");
    } else {
        printf("%s\n",
               "MBD restart initiated");
    }
    return(0);

}

static int
badminDebug (int nargc, char *nargv[], int opCode)
{
    extern char* optarg;

    struct hostInfoEnt *hostInfo;

    char  opt[10];
    char **hostPoint;
    char *word;

    char **hosts=NULL;
    int i, c;
    int send;
    int retCode = 0;
    int  all = FALSE, numHosts = 0;
    struct debugReq  debug;

    debug.opCode = opCode ;
    debug.logClass = 0;
    debug.level = 0;
    debug.hostName = NULL;

    debug.logFileName[0] = '\0';
    debug.options = 0;

    if (opCode == MBD_DEBUG || opCode == SBD_DEBUG)
        strcpy(opt, "oc:l:f:");
    else if (opCode == MBD_TIMING || opCode == SBD_TIMING)
        strcpy (opt, "ol:f:");
    else
        return (-2);

    while ((c = getopt(nargc, nargv, opt )) != EOF) {

        switch (c) {
        case 'c':
            while (optarg != NULL && (word = getNextWord_(&optarg))) {
		if (strcmp(word, "LC_SCHED") == 0)
		    debug.logClass |= LC_SCHED;

		if (strcmp(word, "LC_EXEC") == 0)
		    debug.logClass |= LC_EXEC;

		if (strcmp(word, "LC_TRACE") == 0)
		    debug.logClass |= LC_TRACE;

		if (strcmp(word, "LC_COMM") == 0)
		    debug.logClass |= LC_COMM;

		if (strcmp(word, "LC_XDR") == 0)
		    debug.logClass |= LC_XDR;

		if (strcmp(word, "LC_CHKPNT") == 0)
		    debug.logClass |= LC_CHKPNT;

		if (strcmp(word, "LC_FILE") == 0)
		    debug.logClass |= LC_FILE;

		if (strcmp(word, "LC_AUTH") == 0)
		    debug.logClass |= LC_AUTH;

		if (strcmp(word, "LC_HANG") == 0)
		    debug.logClass |= LC_HANG;

		if (strcmp(word, "LC_SIGNAL") == 0)
		    debug.logClass |= LC_SIGNAL;

		if (strcmp(word, "LC_PIM") == 0)
		    debug.logClass |= LC_PIM;

		if (strcmp(word, "LC_SYS") == 0)
		    debug.logClass |= LC_SYS;

		if (strcmp(word, "LC_JLIMIT") == 0)
		    debug.logClass |= LC_JLIMIT;

		if (strcmp(word, "LC_PEND") == 0)
		    debug.logClass |= LC_PEND;

		if (strcmp(word, "LC_LOADINDX") == 0)
		    debug.logClass |= LC_LOADINDX;

		if (strcmp(word, "LC_M_LOG") == 0) {
		    debug.logClass |= LC_M_LOG;
		}

		if (strcmp(word, "LC_PERFM") == 0) {
		    debug.logClass |= LC_PERFM;
		}

		if (strcmp(word, "LC_MPI") == 0) {
		    debug.logClass |= LC_MPI;
		}

		if (strcmp(word, "LC_JGRP") == 0) {
		    debug.logClass |= LC_JGRP;
		}

	    }
            if (debug.logClass == 0)
            {
		fprintf(stderr,"Command denied.Invalid class name\n");
                return (-1);
            }
            break;

        case 'l':
            for (i=0;i<strlen(optarg);i++) {
		if (!isdigit(optarg[i])) {
                    fprintf(stderr,"Command denied. Invalid level value\n");
		    return(-1);
		}
	    }
	    debug.level = atoi(optarg);
            if (opCode == MBD_DEBUG || opCode == SBD_DEBUG) {
                if (debug.level < 0 || debug.level > 3) {
                    fprintf(stderr,"Command denied. Valid debug level is [0-3] \n");
                    return(-1);
                }
            }
            else if (debug.level < 1 || debug.level > 5) {
                fprintf(stderr,"Command denied. Valid timing level is [1-5]\n");
                return(-1);
	    }
            break;

        case 'f':
	    if (strstr(optarg,"/") && strstr(optarg,"\\")) {
		fprintf(stderr,"Command denied. Invalid file name\n");
                return(-1);
            }
            memset(debug.logFileName,0,sizeof(debug.logFileName));
            ls_strcat(debug.logFileName,sizeof(debug.logFileName),optarg);
            if (debug.logFileName[strlen(debug.logFileName)-1] == '/' ||
                debug.logFileName[strlen(debug.logFileName)-1] == '\\') {
                fprintf(stderr,"Command denied. File name is needed after the path\n");
                return(-1);
            }
	    break;
        case 'o':
 	    debug.options = 1;
	    break;

        default:
	    return (-2);
        }
    }

    if ( opCode == SBD_DEBUG || opCode == SBD_TIMING) {

        numHosts = getNames (nargc, nargv, optind, &hosts, &all, "hostC");
        hostPoint = NULL;
        if (!numHosts && !all)
            numHosts = 1;
        else if (numHosts)
            hostPoint = hosts;

        if ((hostInfo = lsb_hostinfo (hostPoint, &numHosts)) == NULL) {
            lsb_perror(NULL);
            return (-1);
        }

        for (i = 0; i < numHosts; i++) {
            if (strcmp(hostInfo[i].host, "lost_and_found") == 0 ) {
                if (!all)
                    fprintf(stderr, "%s.\n",
                            "<lost_and_found> is not a real host, ignored");
                continue;
            }

            fflush(stderr);
            if (hostInfo[i].hStatus
                & (HOST_STAT_UNAVAIL | HOST_STAT_UNREACH)) {
                if (hostInfo[i].hStatus & HOST_STAT_UNAVAIL)
                    fprintf(stderr, "failed : LSF daemon (LIM) is unavailable on host %s\n",
                            hostInfo[i].host);
                else
		    fprintf(stderr, "failed : Slave batch daemon (sbatchd) is unreachable now on host %s\n",
			    hostInfo[i].host);
                continue;
            }

            if ((send = lsb_debugReq (&debug, hostInfo[i].host)) < 0) {
                char msg[100];
                sprintf(msg,
			"Operation denied by SBD on <%s>",
			hostInfo[i].host);
                lsb_perror(msg);
                retCode = -1;
            }
        }
	return(retCode);
    }

    else {
        numHosts = getNames (nargc, nargv, optind, &hosts, &all, "hostC");
	if (numHosts > 0) {
	    fprintf(stderr, "Host name does not need to be specified, set debug to the host which runs MBD\n");
	}
	if ((send = lsb_debugReq (&debug, NULL)) < 0) {

	    char msg[100];
 	    sprintf(msg, "Operation denied by MBD");
 	    lsb_perror(msg);
	    return (-1);
        }
    }
    return (0);

}
